### PR TITLE
[ECAM] Throttle lever position and AVAIL message on N1 gauge fix

### LINKED ISSUE #1527: [BUG] Upper ECAM thrust lever position not updating
**Mod Version**

"master"

**Describe the bug**

Thrust lever position indicator on upper ecam does not mimic physical thrust lever.

**To Reproduce**

1.Load Aircraft
2.Start Engines
3.Move Thrust Lever and watch at upper ECAM

**Expected behavior**

Thrust lever indication on upper ECAM should match physical thrust lever position.

**Actual behavior**

Thrust lever indication shows no movement until thrust lever was moved about 20%.
See reference Image. Thrust lever indication will start moving above that percentage and will follow the arc smoothly.

**References**

image

In this Image thrust lever was not moved for some time
image

**Additional context**

Was this working before/when did the issue start occurring?
Since #1477 merge
Is this a problem in the vanilla unmodded game?
No.

Discord username (if different from GitHub): DerL30N#3751

### LINKED ISSUE #1596: [BUG] [ECAM] N1 donut behavior not correct
**Mod Version**

Current master as of 29th Okt 20

**Describe the bug**

Behavior of the Thrust lever postion indicator (donut) lags behind and is not correct. 
Donut starting position is already fixed by #1549 

This is the point where the donut starts moving. Already lagging behind. **Levers at 20%**
image

**Thrust set at 50% with levers only at 38%**
image

**Levers 100%**
image

**To Reproduce**

1. Load a flight
2. Look at the indication
3.

**Expected behavior**

donut should match up with the N1 indicator when in manual thrust

**Actual behavior**

**References**

**Additional context**

Is this a problem in the vanilla unmodded game?

Discord username (if different from GitHub):

### LINKED ISSUE #1626: [BUG] AVAIL/REV indication are missing
**Mod Version**

built: 2020-10-30T02:12:21+00:00
ref: refs/heads/master
sha: 772b70172905237d6db8dfe6fcf58701708c15b0
actor: Benjozork
event_name: push

**Describe the bug**

AVAIL message during engine start is gone since last update. I'm using the DEV version (uptodate)
Same for REV (reverse)

**To Reproduce**

1. Star the engine, the new added indication "AVAIL" is now missing on the upper ECAM
2. Use the reverse when engines are on. You'll not see the REV on the upper ECAM
3.

**Expected behavior**

We should see AVAIL or REV.

**Actual behavior**

Nothing appear

**References**

**Additional context**

Was this working before/when did the issue start occurring?

Is this a problem in the vanilla unmodded game?

Discord username (if different from GitHub):

**BY REMOVING THIS LINE, I ATTEST THAT I HAVE FOLLOWED THE ISSUE TEMPLATE AND PROVIDED A TITLE, as well as searched for other possible duplicate issues. I acknowledge if this line is not removed and/or the issue template is not respected, my issue will be closed without warning**

### LINKED ISSUE #1681: [BUG] Wrong appearance of N2
**Mod Version**

latest master

**Describe the bug**

a zero appeared after the XX while engine are off 

**To Reproduce**

1. start the plane on C & D state 
2.
3.

**Expected behavior**

the zero shouldn't be there 

**Actual behavior**

the zero is there 

**References**

Desktop Screenshot 2020 11 01 - 02 52 10 100

**Additional context**

Was this working before/when did the issue start occurring?
the zero wasn't there before 
Is this a problem in the vanilla unmodded game?
yes

Discord username (if different from GitHub):

### LINKED ISSUE #1755: [BUG] Engine 1 information not showing in ECAM
**Mod Version**
Dev Version

**Describe the bug**
Engine 1 informations are not showing on the ECAM (N1, EGT, N2, FF). xxx remains even after the engine is running correctly

**To Reproduce**
1. turn on engine 2 
2. turn on engine 1

**Expected behavior**
N1, EGT, N2, FF should increase according on the engine start up

**Actual behavior**
xxx remains even if the engine is running

**References**

**Additional context**

Was this working before/when did the issue start occurring?

It was working with the dev mod of 30th oct

Is this a problem in the vanilla unmodded game?

No

### PR DESCRIPTION
**Pull request by RichardPilbery**
Fixes #1626, Fixes #1596, Fixes #1527, Fixes #1681, Fixes #1755

## Summary of Changes

Sigh. N1 on upper ECAM is a problem. The values provided by simvar value ENG N1 RPM seem to be consistently higher than the thrust set below 80-90%. I've had to create a convoluted set of rules to either return this value or the throttle position percentage based on the N1 and throttle values. It does mean the cursor 'sticks' to the donut, which is not ideal. Not sure what else can be done unless there are accurate N1 values coming from the sim.

In better news, the AVAIL and REV messages now appear
I've also added an semi-opaque background to the N2 value when the engine is starting.

## Screenshots (if necessary)

avail

Highlight-N2

50percent

## References

The issue reports refer to several. Also see below:

https://youtu.be/4NAs7DJ42eM?t=114

## Additional context

Discord username (if different from GitHub): tricky_dicky#3571

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build.py will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Comment by MisterChocker** [comment]
Please write:
Fixes issue#
Fixes issue#
Fixes issue#

otherwise issues won't be automatically closed.

**Comment by lukecologne** [comment]
I've updated the comment for you @RichardPilbery

**Comment by lukecologne** [comment]
> The values provided by simvar value ENG N1 RPM seem to be consistently higher than the thrust set below 80-90%

Have you tried the corrected N1? Simvar for that is TURB ENG CORRECTED N1:index

**Comment by RichardPilbery** [comment]
Thanks @lukecologne for amending the comment and your suggestions. Based on the values, it looks like ENG N1 RPM does provide the percentage, but I'll take another look at the variables you suggested.

**Comment by z0rgmeister** [comment]
Name: z0rgmeister#8972
Date of testing: 02/11/2020
Version of the sim: 1.10.8.0
PR Tested: #1650
Object of test: Test N1% and Donut behaviour in relation to physical thrust levers

Steps taken: Single straight from runway flight at LPPT loaded with default loadout. 

Flight notes: NA

Taxi/Takeoff : NA

Cruise: NA

Descent/Approach/Landing: NA

Shutdown: NA

Issues: Regressing movement maybe too quick on the N1%.
Overall Rating: Good
Conclusions of the testing: Good spooling up and smooth movement of the donut in relation to the levers. When spooling down N1% spool downfeels too quick. Not an Airbus expert though so would leave it to someone who knows to chime in on that.

**Comment by MisterChocker** [comment]
@z0rgmeister okay the engine No. 1 and No. 2 sounds important to me, but the spooling times not covered with this pr.

**Comment by RichardPilbery** [comment]
@lukecologne Sadly ENG N1 RPM (when set as a percentage), TURB ENG N1 and TURB ENG N1 CORRECTED provide almost exactly the same values.

**Comment by MisterChocker** [comment]
Name: @MisterChocker 
Date of testing: Current date 03/11/2020
Version of the sim: 1.10.8.0
PR Tested: #1650
Object of test: Check for power logic, check donut behavior, responses and functionality

## Testing technique:
At EDDT in Cold & Dark, engine start/engine shutdown whilst maintaining parking position, parking break engaged.

## Steps you've made to test the Pull Request:

- Spawned C&D at EDDT on stand.

- Powered aircraft via GPU, tested systems.
ALL GOOD.
- Monitored FADEC self test and power logic, waited for FADEC to be turned off.
ALL GOOD.
image
image

- Set Engine Mode Selector to IGNITION START.
  - FADEC gets powered as expected.
  - Engine parameter are displayed, except for EGT TEMP, still XX readout with green color.
image

- Set Engine Mode Selector to NORM, waited for FADEC to be turned off.
ALL GOOD.
image

- Set Engine Mode Selector to IGNITION START.
  - FADEC gets powered as expected.
Engine parameter are displayed, except for EGT TEMP, still XX readout with green color.
  - Moved Thrust Lever until movement on Donut can be seen.
Donut starts moving ~20% N1.
image
  - Moved Thrust Lever until donut shows movement
image
  - Moved Thrust Lever asynchronous
left and right are independent
image
image

- Set ENGINE MODE SELECTOR to NORM, started APU
when AVAIL opened APU BLEED



- Continued with Engine Start
EGT TEMP shows XX, however changes as soon as EGT changes.
image
lowest value displayed was 15°
image
two good engine starts
image

- Moved Thrust Lever 
experienced some very unnatural behavior.
  - The N1% readout of the engines N1% was not correct.
Please watch [video](https://youtu.be/wNiPyc9hzY8), describing this is extremely difficult.

- Moved Thrust Lever independently
image
image

- Proceeded with engine shutdown
N1% readout seems to be normal
image

- Engine are now off
image

- FADEC power supply cut off
image

- Engine Mode Selector set to IGNITION START, FADEC get powered again
image


## Flight notes

- Startup: 
  - When FACED get unpowered and powered again, Engine EGT TEMP readout is XX in green.
  - Odd behavior on engine N1% readout when moving Thrust Lever

- Taxi/Takeoff : N/A

- Cruise: N/A

- Descent/Landing : N/A

- Shutdown: N/A

## Issues (If Any):
Overall Rating: Needs improvements.

Conclusions of the testing:
- FADEC power logic on EGT needs to be corrected.
- Donut will show no movement below ~20% N1.
- Engine N1% readout represents not the correct engine N1%. (Almost if engine N1% readouts are not allowed to be higher then the representative Thrust Lever position.

**Comment by tareksabet** [comment]
The XX on the N2 should  be aligned with the XX on FuelFlow

**Comment by RichardPilbery** [comment]
Thanks @MisterChocker for the comprehensive review. You're right that the logic currently limits N1 to the throttle position. Thinking about it, I've got this the wrong way round. The throttle position should be based on N1 values. I'll mark this PR as draft and rework the logic.

**Comment by MisterChocker** [comment]
> The XX on the N2 should be aligned with the XX on FuelFlow

The reason why could be because the one text is aligned right and the other isn't.

Real Life:
image

**Comment by RichardPilbery** [comment]
@MisterChocker I think it's  because N2 now has separate text elements to allow for styling of the decimal point and decimal value, so the 'whole number' (which in this case will be set to XX) has been offset. This shouldn't be too tricky to fix.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js [hunk 1]
```diff
@@ -1597,29 +1597,15 @@ var A320_Neo_UpperECAM;
...
+            return value;
         }
         getN1GaugeThrottleValue() {
             const throttle = Math.abs(Simplane.getEngineThrottleCommandedN1(this.index));
             const value = throttle * A320_Neo_UpperECAM.Definitions.THROTTLE_TO_N1_GAUGE;
             return value;
         }
         getThrottlePosition() {
-            return Simplane.getEngineThrottle(this.index);
+            return Math.abs(Simplane.getEngineThrottleCommandedN1(this.index));
```
Comment (MisterChocker) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:1608]: Why abs? can this value get negative?
May use `.toFixed(0)` to round.

Comment (lukecologne) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:1608]: Why would you use toFixed to round? As far as I know math.round is way faster, and more expressive to what you are doing

**Comment by MisterChocker** [comment]
Name: @MisterChocker 
Date of testing: Current date 04/11/2020
Version of the sim: 1.10.8.0
PR Tested: #1650
Object of test: Check for power logic, check donut behavior, responses and functionality

## Testing technique:
At EDDT in Cold & Dark, engine start/engine shutdown whilst maintaining parking position, parking break engaged.

## Steps you've made to test the Pull Request:

- Spawned C&D at EDDT on stand.

- Powered aircraft via GPU, tested systems.
ALL GOOD.
- Monitored FADEC self test and power logic, waited for FADEC to be turned off.
ALL GOOD.
image
image

- Set Engine Mode Selector to IGNITION START.
  - FADEC gets powered as expected.
  - Engine parameter are displayed, except for EGT TEMP, still XX readout with green color.
image

- Set Engine Mode Selector to NORM, waited for FADEC to be turned off.
ALL GOOD.
image

- Set Engine Mode Selector to IGNITION START.
  - FADEC gets powered as expected.
Engine parameter are displayed, except for EGT TEMP, still XX readout with green color.
  - Moved Thrust Lever until movement on Donut can be seen.
Donut starts moving instantly an properly. (You have to look closely)
image
  - Moved Thrust Lever asynchronous
left and right are independent.

- Set ENGINE MODE SELECTOR to NORM, started APU
when AVAIL opened APU BLEED

- Continued with Engine Start
EGT TEMP shows XX, however changes as soon as EGT changes.
two good engine starts
image

- Moved Thrust Lever 
  - Engine No.2 does not react to thrust lever movements
image

  - Engine No.1 has good and natural response to thrust lever movements, however engine No.2 also reacts to engine No.1 thrust lever
image

  - Engine N1% now represents engines N1%, can be seen when moving thrust lever faster.
image
image

- Proceeded with engine shutdown
ALL GOOD

- Engines are now off
image

- FADEC power supply cut off
image

- Engine Mode Selector set to IGNITION START, FADEC get powered again
image


## Flight notes

- Startup: 
  - When FACED get unpowered and powered again, Engine EGT TEMP readout is XX in green.
  - Perfect donut movement on thrust lever movements.
  - Engine No.2 thrust lever has no influence on engine No.2
  - Good engine N1% readout and response
  - Good alignment of engine N1% and donut

- Taxi/Takeoff : N/A

- Cruise: N/A

- Descent/Landing : N/A

- Shutdown: N/A

## Issues (If Any):
Overall Rating: Needs improvements.

Conclusions of the testing:
- FADEC power logic on EGT needs to be corrected.
- Engine No.2 thrust lever has no command over engine No.2
- Engine No.1 thrust lever controls engine No.1 and No.2

**Comment by MisterChocker** [comment]
When FADEC off, I don't see boxes around EGT TEMP, this could also be added. Should actually be pretty simple, just an if else statement.

> Real Life:
> image

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js [hunk 2]
```diff
@@ -1779,6 +1816,13 @@ var A320_Neo_UpperECAM;
             const percent = SimVar.GetSimVarValue(name, "percent");
             return percent;
         }
+        getDisplayActiveEngine(_engine) {
+            if (this.getValue(_engine) < 57.8 && this.getEngineStartStatus(_engine) && this.getIgnitionStatus(_engine)) {
```
Comment (MisterChocker) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:1820]: Simplane.getEngineActive(0) || Simplane.getEngineActive(1)
returns true if engine has N2 > 20% (could be less, not sure)

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js (hunk 1 above)
Comment (MisterChocker) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:1608]: dunno, new to js, have seen it all over the place, but abs != round

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js (hunk 2 above)
Comment (RichardPilbery) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:1820]: Not sure what advantage that conveys over the proposed code change.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js (hunk 1 above)
Comment (RichardPilbery) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:1608]: I suspect the throttle value goes negative on reverse thrust but would need to check. This is original Asobo code I think.

**Comment by RichardPilbery** [comment]
Actions:
- [ ] FADEC power logic on EGT needs to be corrected @RichardPilbery 
- [ ] Remove box from around EGT when engine inactive @RichardPilbery 
- [ ] Create issue relating to throttle behaviour as this is unrelated to this PR: @MisterChocker one for you perhaps?

**Comment by MisterChocker** [comment]
@RichardPilbery 
> * [ ]  Create issue relating to throttle behaviour as this is unrelated to this PR: @MisterChocker one for you perhaps?

I would like to wait for a couple more reviews, I am pretty sure others with hotas (2 levers) will have the same problem.
I have tested different thrust levers with 2 axis and same result for all.

**Comment by Benjozork** [comment]
Please provide a more descriptive title. Thanks !

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_ECAMGauge.js [hunk 3]
```diff
@@ -352,7 +354,7 @@ var A320_Neo_ECAM_Common;
             }
             if (this.currentValueText != null) {
                 this.currentValueText.setAttribute("class", style);
-                if (this.currentValuePrecision == 0) {
+                if (this.uppercam == true) {
```
Comment (MisterChocker) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_ECAMGauge.js:357]: You don't need to check for `true` wit `==` true, even tho `===` would be "more correct"
```js
if (this.upperecam) {
```

**Comment by RichardPilbery** [comment]
> @RichardPilbery
> 
> > * [ ]   Create issue relating to throttle behaviour as this is unrelated to this PR: @MisterChocker one for you perhaps?
> 
> Doesn't this pr shall solve this?

Sometimes a video is easier: https://www.youtube.com/watch?v=rUZ6YOZ8W7k
I cannot replicate your issue with the throttle levers not behaving as expected:

+ Engine No.2 thrust lever has no command over engine No.2
+ Engine No.1 thrust lever controls engine No.1 and No.2

Could it be an issue with your hardware? As the video shows, the throttle levers seem to work independently as expected. I did replicate the steps you outlined above.

**Review by MisterChocker — CHANGES_REQUESTED** [review]

**Comment by MisterChocker** [comment]
> Could it be an issue with your hardware? As the video shows, the throttle levers seem to work independently as expected. I did replicate the steps you outlined above.

Already thought that you would not have the problem. I tested with different hardware, however the results were the same.
Do you have thrust lever axis defined? I don't move the thrust lever with my mouse.

**Comment by RichardPilbery** [comment]
@MisterChocker I have a Thrustmaster throttle, but it does not allow for controlling of engine throttles independently. The only was I can test them is to do so in sim with the mouse. Could be an Asobo bug maybe?

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_ECAMGauge.js (hunk 3 above)
Comment (RichardPilbery) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_ECAMGauge.js:357]: That's true!

**Comment by MisterChocker** [comment]
> Could be an Asobo bug maybe?

I don't know, I was able to use both axis before.

**Comment by tareksabet** [comment]
Name: tarek_sabet#7946
Date of testing: Current date 06/11/2020
Version of the sim: 1.10.8.0
PR Tested: #1650 
Object of test: Upper ECAM and how the N1 works 

Testing technique: Full flight from HECA to HESH
Steps you've made to test the Pull Request:
1. spawn the plane in C & D
2. Do the full flight till shutdown

Flight notes

Startup: everything works as described 

Taxi/Takeoff : everything works as described 

Cruise: everything works as described 

Descent/Landing : everything works as described 

Shutdown: everything works as described 

Issues (If Any): N/A
Overall Rating: Good
Conclusions of the testing: Great job, LGTM

**Comment by starvingjavi** [comment]
Name: Javi#2557
Date of testing: Current date 05/11/2020
Version of the sim: 1.10.8.0
PR Tested: #1650
Object of test: Upper ECAM and N1 gauges

Testing technique: Full flight SUMU to SABE
Steps you've made to test the Pull Request:

1. Spawn the plane in SUMU C&D
2. Continue flight until landing and shutdown
3. Notice ECAM and N1 indicators

Flight notes

Startup: No problems

Taxi/Takeoff : No problems

Cruise: No problems

Descent/Landing : No problems

Shutdown: No problems

Issues (If Any): No problems
Overall Rating: Good
Conclusions of the testing: Works as described

**Pull request merged**